### PR TITLE
De-XHR-ify

### PR DESCRIPTION
**Pull request by wbamberg**
This PR is part of https://github.com/mdn/content/issues/30131, specifically the part where we use fetch() instead of XHR for the examples for the Web Audio API. Since it's complicated, I'm going to try to explain what I'm doing in a bit of detail.

Most of the pages in the Web Audio docs include a code sample that's derived from a full example in webaudio-examples. The table below lists, in the first column, every page that has an example using XHR, and in the right column, which webaudio-examples example it is derived from:

| Page                    | Example                                            |
| ----------------------- | -------------------------------------------------- |
| AudioBufferSourceNode   |                                                    |
| - loop                  | -> decode-audio-promise                            |
| - loopstart             | -> decode-audio-promise                            |
| - loopend               | -> decode-audio-promise                            |
| - playbackrate          | -> decode-audio-promise                            |
| AudioProcessingEvent    | -> script-processor-node                           |
| BaseAudioContext        |                                                    |
| - createConvolver       | -> voice-change-o-matic                            |
| - createScriptProcessor | -> script-processor-node                           |
| - decodeAudioData       | -> decode-audio-promise, decode-audio-data         |
| ConvolverNode           |                                                    |
| - buffer                | -> (no external example)                           |
| - normalize             | -> (no external example)                           |
| OfflineAudioContext     | -> offline-audio-context-promise (but XHR in page) |
| - startRendering        | -> offline-audio-context-promise (but XHR in page) |

Note that:

- the "decode-audio-data" example recently got a extra version, "decode-audio-promise". The main idea there was to show the version of `BaseAudioContext.decodeAudioData()` that returns a promise. We should have both versions, since the callback-based version of the API is still legitimate. However, only the new "decode-audio-promise" uses `fetch()`, the other still uses XHR. They should both use `fetch()`.

- the "offline-audio-context-promise" got an update to use `fetch()` as well, but it seems like the example in the page still quotes the XHR version.

- some of the examples in the code are very cut-down versions of the full examples. It would be nice if they were as complete as possible, but of course the downside there is that the examples will be longer (certainly with something like voice-o-matic it's not helpful to quote the whole code just to show `Convolver.buffer`).

- many of the examples include the JS inline in the HTML. I think this makes them harder to read than separate JS files.

So here's the plan:

- https://github.com/mdn/webaudio-examples/pull/116/commits/4b2ffe92b6fbce5bf6117135aea941ab0b673597: create two new examples, "audio-buffer-source-node/loop" and "audio-buffer-source-node/playbackrate", which effectively split "decode-audio-promise", so the `AudioBufferSourceNode` pages can get more focused examples.
- https://github.com/mdn/webaudio-examples/pull/116/commits/30666524f66de40977899af13cfe2bd95a586558: rename "decode-audio-promise" to "decode-audio-data/promise", and rename "decode-audio-data" to "decode-audio-data/callback", to reflect better the difference between these two examples.
- https://github.com/mdn/webaudio-examples/pull/116/commits/45694ed08c1ad163f928c4388c15084fdc33d649: update all these examples to have JS in separate files.
- https://github.com/mdn/webaudio-examples/pull/116/commits/344e8fb994e1334f5c2951a118ea0eb7f7aa97e5, https://github.com/mdn/webaudio-examples/pull/116/commits/af09b5afd78c4903a786f86f037d78c586b0aa04, https://github.com/mdn/webaudio-examples/pull/116/commits/1461be68f0ff1e497a7d846185c929aa0a0ecc3f: update all these examples to use `fetch()`.
- In the pages, update example code blocks in the page to be more complete, to the extent that this doesn't make them too long. I think we'll keep `Convolver.buffer` and `Convolver.normalize` short at this point though.

Also the voice-change-o-matic example includes a script "respond.js" which seems to be a polyfill for `window.matchMedia()` which is (a) not used and (b) not needed any more, so I deleted it: https://github.com/mdn/webaudio-examples/pull/116/commits/aab9a15f2e4536dd840ba615ce3ce3e445ce4402.

### CODE REVIEW

File: voice-change-o-matic/scripts/app.js [hunk 1]
```diff
@@ -76,34 +76,19 @@ function init() {
...
+    const decodedAudio = await audioCtx.decodeAudioData(arrayBuffer);
+    convolver.buffer = decodedAudio;
+  } catch (error) {
+    console.error(
+      `Unable to fetch the audio file: ${name} Error: ${err.message}`
+    );
+  }
 
   // Set up canvas context for visualizer
   const canvas = document.querySelector(".visualizer");
```
Comment (teoli2003) [inline voice-change-o-matic/scripts/app.js:94]: Line 198 + 202: Remove useless `let barHeight;` and replace `barHeight = dataArrayAlt[i];` with `const barHeight = dataArrayAlt[i];`

File: voice-change-o-matic/scripts/app.js [hunk 2]
```diff
@@ -307,26 +292,24 @@ function init() {
   }
```
Comment (teoli2003) [inline voice-change-o-matic/scripts/app.js:292]: Line 271, 274, 277, 280: use arrow functions.

**Review by teoli2003 — CHANGES_REQUESTED** [review]
Two groups of comments:
- a few NITs
- some more syntax improvements (in voice-of-matic): we are touching these files so rarely (as some are copied in snippets in mdn/content), that it be a good thing to modernize the syntax at the same time.

Really great changes!

File: voice-change-o-matic/scripts/app.js (hunk 2 above)
Comment (wbamberg) [reply voice-change-o-matic/scripts/app.js:292]: I think we can use method shorthand here even:

```js
    return {
      apply() {
        wetNode.gain.setValueAtTime(0.75, audioContext.currentTime);
      },
      discard() {
        wetNode.gain.setValueAtTime(0, audioContext.currentTime);
      },
      isApplied() {
        return wetNode.gain.value > 0;
      },
      placeBetween(inputNode, outputNode) {
        inputNode.connect(delay);
        delay.connect(wetNode);
        wetNode.connect(filter);
        filter.connect(delay);

        inputNode.connect(dryNode);
        dryNode.connect(mixer);
        wetNode.connect(mixer);
        mixer.connect(outputNode);
      },
```

**Comment by wbamberg** [comment]
> some more syntax improvements (in voice-of-matic): we are touching these files so rarely (as some are copied in snippets in mdn/content), that it be a good thing to modernize the syntax at the same time.

Yeah, fair enough. I wasn't sure how far to go with unrelated changes (and I was just getting tired of it last night) but it does make sense to do these updates as well, while we are here.

File: audio-buffer-source-node/loop/script.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
+const audioCtx = new AudioContext();
```
Comment (teoli2003) [inline audio-buffer-source-node/loop/script.js:1]: This raises an error in Chrome (only) that doesn't allow the creation of `AudioContext` object before a user gesture on the page. (Security requirement: The _Autoplay policy in Chrome_ https://goo.gl/7K7WLu ).
Capture d’écran 2023-11-16 à 10 02 59

The example does work, but we should avoid this. (I had to workaround around this the last time I updated some examples here)

Comment (teoli2003) [reply audio-buffer-source-node/loop/script.js:1]: I've updated the script in the next comment (selecting the whole script.js file)

Question: do you want to update the other scripts (several have the same problem) or do you want to do ti yourself?

File: audio-buffer-source-node/loop/script.js [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  source.loopStart = loopstartControl.value;
+  loopstartValue.textContent = loopstartControl.value;
+});
+
+loopendControl.addEventListener("input", () => {
+  source.loopEnd = loopendControl.value;
+  loopendValue.textContent = loopendControl.value;
+});
+
+loadAudio();
```
Comment (teoli2003) [inline audio-buffer-source-node/loop/script.js:63]:
```suggestion
let audioCtx;
let buffer;
let source;

const play = document.getElementById("play");
const stop = document.getElementById("stop");

const loopstartControl = document.getElementById("loopstart-control");
const loopstartValue = document.getElementById("loopstart-value");

const loopendControl = document.getElementById("loopend-control");
const loopendValue = document.getElementById("loopend-value");

async function fetchAudio(name) {
  try {
    const response = await fetch(name);
    return audioCtx.decodeAudioData(await response.arrayBuffer());
  } catch (err) {
    console.error(
      `Unable to fetch the audio file: ${name} Error: ${err.message}`
    );
  }
}

async function loadAudio() {
  buffer = await fetchAudio("rnb-lofi-melody-loop.wav");
  const max = Math.floor(buffer.duration);
  loopstartControl.setAttribute("max", max);
  loopendControl.setAttribute("max", max);
}

play.addEventListener("click", async () => {
  if (!audioCtx) {
    audioCtx = new AudioContext();
    await loadAudio();
  }
  source = audioCtx.createBufferSource();
  source.buffer = buffer;
  source.connect(audioCtx.destination);
  source.loop = true;
  source.loopStart = loopstartControl.value;
  source.loopEnd = loopendControl.value;
  source.start();
  play.disabled = true;
  stop.disabled = false;
  loopstartControl.disabled = false;
  loopendControl.disabled = false;
});

stop.addEventListener("click", () => {
  source.stop();
  play.disabled = false;
  stop.disabled = true;
  loopstartControl.disabled = true;
  loopendControl.disabled = true;
});

loopstartControl.addEventListener("input", () => {
  source.loopStart = loopstartControl.value;
  loopstartValue.textContent = loopstartControl.value;
});

loopendControl.addEventListener("input", () => {
  source.loopEnd = loopendControl.value;
  loopendValue.textContent = loopendControl.value;
});

```

File: audio-buffer-source-node/loop/index.html [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    <title>Web Audio API examples: looping a track</title>
+
+    <script src="script.js" defer></script>
+  </head>
+
+  <body>
+    <h1>Web Audio API examples: looping a track</h1>
+
+    <button id="play" disabled>Play</button>
+    <button id="stop">Stop</button>
```
Comment (teoli2003) [inline audio-buffer-source-node/loop/index.html:16]: Small change related to the change in where AudioContext is created in the script.js.
```suggestion
    <button id="play>Play</button>
    <button id="stop" disabled>Stop</button>
```

File: audio-buffer-source-node/playbackrate/index.html [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    <h2>Set playback rate</h2>
+    <input
+      id="playback-rate-control"
+      type="range"
+      min="0.25"
+      max="3"
+      step="0.05"
+      value="1"
+      disabled />
+    <span id="playback-rate-value">1.0</span>
```
Comment (teoli2003) [inline audio-buffer-source-node/playbackrate/index.html:27]:
```suggestion
    <span id="playback-rate-value">1.0</span>  × the original rate.
```

File: audio-buffer-source-node/playbackrate/script.js [hunk 7]
```diff
@@ -0,0 +1,49 @@
+const audioCtx = new AudioContext();
```
Comment (teoli2003) [inline audio-buffer-source-node/playbackrate/script.js:1]: For this example, I broke down in small suggestions the changes I did for working around the _Autoplay policy in Chrome_.
```suggestion
let audioCtx;
```

File: audio-buffer-source-node/playbackrate/script.js [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    );
+  }
+}
+
+async function loadAudio() {
+  buffer = await fetchAudio("rnb-lofi-melody-loop.wav");
+  play.disabled = false;
+}
+
+play.addEventListener("click", () => {
```
Comment (teoli2003) [inline audio-buffer-source-node/playbackrate/script.js:27]:
```suggestion
play.addEventListener("click", async () => {
  if (!audioCtx) {
    audioCtx = new AudioContext();
    await loadAudio();
  }
```

File: audio-buffer-source-node/playbackrate/script.js [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+  source = audioCtx.createBufferSource();
+  source.buffer = buffer;
+  source.connect(audioCtx.destination);
+  source.loop = true;
+  source.playbackRate.value = playbackControl.value;
+  source.start();
+  play.disabled = true;
```
Comment (teoli2003) [inline audio-buffer-source-node/playbackrate/script.js:34]:
```suggestion
  play.disabled = true;
  stop.disabled = false;
```

File: audio-buffer-source-node/playbackrate/script.js [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+  playbackControl.disabled = false;
+});
+
+stop.addEventListener("click", () => {
+  source.stop();
+  play.disabled = false;
```
Comment (teoli2003) [inline audio-buffer-source-node/playbackrate/script.js:40]:
```suggestion
  play.disabled = false;
  stop.disabled = true;
```

File: audio-buffer-source-node/playbackrate/script.js [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+  playbackControl.disabled = true;
+});
+
+playbackControl.oninput = () => {
+  source.playbackRate.value = playbackControl.value;
+  playbackValue.textContent = playbackControl.value;
+};
+
+loadAudio();
```
Comment (teoli2003) [inline audio-buffer-source-node/playbackrate/script.js:49]:
```suggestion
```

File: audio-buffer-source-node/playbackrate/index.html [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+    <title>Web Audio API examples: setting playback rate</title>
+
+    <script src="script.js" defer></script>
+  </head>
+
+  <body>
+    <h1>Web Audio API examples: setting playback rate</h1>
+
+    <button id="play" disabled>Play</button>
+    <button id="stop">Stop</button>
```
Comment (teoli2003) [inline audio-buffer-source-node/playbackrate/index.html:16]:
```suggestion
    <button id="play">Play</button>
    <button id="stop" disabled>Stop</button>
```

**Review by teoli2003 — CHANGES_REQUESTED** [review]
The AudioContext creation "problem" with the _Autoplay policy in Chrome_ is the major thing. (See comments inline)

I've updated the _loop_ and _playback_ examples here (among some nits). I can do the others if you want (after your review of my proposed changes), or you can do them (as you want))

**Comment by wbamberg** [comment]
> The AudioContext creation "problem" with the _Autoplay policy in Chrome_ is the major thing. (See comments inline)
> 
> I've updated the _loop_ and _playback_ examples here (among some nits). I can do the others if you want (after your review of my proposed changes), or you can do them (as you want))

All right, I had a go at implementing this in all of the relevant examples, AFAICT.

I made some slightly unrelated changes:
- use a `<p>` for the note instead of `<br>` elements
- merge `fetchAudio` into `loadAudio`
- consistently use `addEventListener`

The "callback" version of decodeAudioData needed some slightly different handling, but it should be reasonably clear.

**Review by teoli2003 — APPROVED** [review]
👍

**Pull request merged**